python/tskit/combinatorics.py

```diff
@@ -24,13 +24,244 @@
+        # Avoid recomputing anything for the parent until all child edges
+        # for that parent are inserted/removed
+        for p, sibling_edges in itertools.groupby(edges_out, key=lambda e: e.parent):
+            for e in sibling_edges:
+                parent[e.child] = -1
+            update_state(tree, p)
+        for p, sibling_edges in itertools.groupby(edges_in, key=lambda e: e.parent):
+            if tree.is_sample(p):
+                raise ValueError("Internal samples not supported.")
+            for e in sibling_edges:
+                parent[e.child] = p
+            update_state(tree, p)
+
+        counters = []
+        for root in tree.roots:
+            if topology_counter[root] is not None:
+                counters.append(topology_counter[root])
+        yield TopologyCounter.merge(counters)
+
+
+def tree_count_topologies(tree, sample_sets):
+    for u in tree.samples():
+        if not tree.is_leaf(u):
+            raise ValueError("Internal samples not supported.")
+
+    topology_counter = np.full(tree.tree_sequence.num_nodes, None, dtype=object)
+    for sample_set_index, sample_set in enumerate(sample_sets):
+        for u in sample_set:
+            if not tree.is_sample(u):
+                raise ValueError(f"Node {u} in sample_sets is not a sample.")
+            topology_counter[u] = TopologyCounter.from_sample(sample_set_index)
+
+    for u in tree.nodes(order="postorder"):
+        children = []
+        for v in tree.children(u):
+            if topology_counter[v] is not None:
+                children.append(topology_counter[v])
+        if len(children) > 0:
+            topology_counter[u] = combine_child_topologies(children)
+
+    counters = []
+    for root in tree.roots:
+        if topology_counter[root] is not None:
+            counters.append(topology_counter[root])
+    return TopologyCounter.merge(counters)
+
+
+def combine_child_topologies(topology_counters):
+    """
+    Select all combinations of topologies from different
+    counters in ``topology_counters`` that are capable of
+    being combined into a single topology. This includes
+    any combination of at least two topologies, all from
+    different children, where no topologies share a
+    sample set index.
+    """
+    partial_topologies = PartialTopologyCounter()
+    for tc in topology_counters:
+        partial_topologies.add_sibling_topologies(tc)
+
+    return partial_topologies.join_all_combinations()
+
+
+class TopologyCounter:
+    """
+    Contains the distributions of embedded topologies for every combination
+    of the sample sets used to generate the ``TopologyCounter``. It is
+    indexable by a combination of sample set indexes and returns a
+    ``collections.Counter`` whose keys are topology ranks
+    (see :ref:`sec_tree_ranks`), and whose counts represent the probability
+    of that topology if one randomly selected an sample from each
+    sample set pertaining to that ``Counter``.
```

Comment:
the last part of this sentence doesn't quite make sense to me?

Reply:
After reading the documentation for `count_topologies` I know what this does - maybe just refer to that method?